src/client/scripts/game/chess/checkmatepractice.js

```diff
@@ -0,0 +1,123 @@
+        if (!validCheckmates.includes(checkmateID)) return console.error("User tried to play invalid checkmate practice.");
+
+        let startingPosition = {}; // the position to be generated
+        let blackpieceplaced = false; // monitors if a black piece has already been placed
+        let whitebishopparity = Math.floor(Math.random() * 2); // square color of first white bishop batch
+        
+        // read the elementID and convert it to a position
+        const piecelist = checkmateID.match(/[0-9]+[a-zA-Z]+/g);
+        for (let entry of piecelist) {
+            let amount = entry.match(/[0-9]+/)[0]; // number of pieces to be placed
+            let piece = entry.match(/[a-zA-Z]+/)[0]; // piecetype to be placed
+            piece = formatconverter.ShortToLong_Piece(piece);
+
+            // place amount many pieces of type piece
+            while (amount != 0) {
+                if (math.getPieceColorFromType(piece) === "white") {
+                    if (blackpieceplaced) return console.error("Must place all white pieces before placing black pieces.");
+
+                    // randomly generate white piece coordinates near origin in square from -5 to 5
+                    const x = Math.floor(Math.random() * 11) - 5;
+                    const y = Math.floor(Math.random() * 11) - 5;
+                    const key = math.getKeyFromCoords([x,y]);
+
+                    // check if square is occupied and white bishop parity is fulfilled
+                    if (!(key in startingPosition) && !(piece == "bishopsW" && (x + y)%2 != whitebishopparity)) {
+                        startingPosition[key] = piece;
+                        amount -= 1;
+                    }
+                } else {
+                    // randomly generate black piece coordinates at a distance
+                    const x = Math.floor(Math.random() * 3) + 12;
+                    const y = Math.floor(Math.random() * 35) - 17;
+                    const key = math.getKeyFromCoords([x,y]);
+                    // check if square is occupied or potentially threatened
+                    if (!(key in startingPosition) && squareNotInSight(key, startingPosition)) {
+                        startingPosition[key] = piece;
+                        amount -= 1;
+                        blackpieceplaced = true;
+                    }
+                }
+            }
+
+            // flip white bishop parity
+            whitebishopparity = 1 - whitebishopparity;
+        }
+
+        return startingPosition;
+    }
+
+    /**
+     * This method checks that the input square is not on the same row, column or diagonal as any key in the startingPosition object
+     * It also checks that it is not attacked by a knightrider
+     * @param {string} square - square of black piece
+     * @param {Object} startingPosition - startingPosition JSON containing all white pieces
+     * @returns true or false, depending on if the square is in sight or not
```

Comment:
* @returns {boolean} true or false, depending on if the square is in sight or not